guides/ci/travis.md

```diff
@@ -0,0 +1,252 @@
+
+And here is the Travis CI configuration file:
+
+```yml
+# .travis.yml
+language: bash
+
+services:
+  - docker
+
+before_install:
+  # build image using Dockerfile:
+  - docker build -t testing .
+
+script:
+  # run specs in the container
+  - docker run -v $PWD:/src -w /src testing crystal spec
+```
+
+**Note:** Dockerfile arguments can be used to use the same Dockerfile for latest, nightly or a specific version.
+
+## Using a database (example using MySQL)
+
+As you may notice, in the last configuration file we are using `docker` as a `service`. Travis CI may start different services as needed. For example, we may need [MySQL](https://docs.travis-ci.com/user/database-setup/#mysql).
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+services:
+  - mysql
+
+script:
+  - crystal spec
+```
+
+Here is the new test file for testing against the database:
+
+```crystal
+# spec/simple_db_spec.cr
+require "./spec_helper"
+require "mysql"
+
+it "connects to the database" do
+  DB.connect ENV["DATABASE_URL"] do |cnn|
+    cnn.query_one("SELECT 'foo'", as: String).should eq "foo"
+  end
+end
+```
+
+When pushing this changes Travis CI will report the following error: `Unknown database 'test' (Exception)`, showing that we need to configure the MySQL service **and also setup the database**:
+
+> It's really **important** to notice that the lines we are adding to `.travis.yml` will depend exclusively on the development workflow we are using!
+> And remember that this is only an example using MySQL.
+
+```yaml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+env:
+  - DATABASE_URL="mysql://root@localhost/test"
+
+services:
+  - mysql
+
+before_install:
+  - mysql -u root --password="" < test-data/setup.sql
+
+script:
+  - crystal spec
+```
+
+We are [using a `setup.sql` script](https://andidittrich.de/2017/06/travisci-setup-mysql-tablesdata-before-running-tests.html) to create a more readable `.travis.yml`. The file `./test-data/setup.sql` looks like this:
+
+```sql
+-- setup.sql
+CREATE DATABASE IF NOT EXISTS test;
+-- CREATE TABLE ... etc ...
+```
+
+Pushing these changes will trigger Travis CI and the build should be successful!!
+
+## Caching
+
+If we read Travis CI job log, we will find that every time the job runs, Travis CI needs to fetch the libraries needed to run the application:
+
+```log
+Fetching https://github.com/crystal-lang/crystal-mysql.git
+Fetching https://github.com/crystal-lang/crystal-db.git
+```
+
+This takes time and on the other hand these libraries might not change as often as our application, so it looks like we may cache them and save time.
+
+Travis CI [uses caching](https://docs.travis-ci.com/user/caching/) to improve some parts of the building path. For our example. Here is the new configuration file **with cache enabled**:
+
+```yml
+# .travis.yml
+language: crystal
+crystal:
+  - latest
+
+env:
+  - DATABASE_URL="mysql://root@localhost/test"
+
+cache:
+  directories:
+    - $HOME/.cache/shards
```

Comment:
`$HOME/.cache/shards` caches the git repos, but they still need to be checked out on every run. `$TRAVIS_BUILD_DIR/lib` should be cached as well. Or even only `$TRAVIS_BUILD_DIR/lib`.

Reply:
I disagree. If the local lib directory is cached then it will not mimic a fresh checkout of the repo. The caching of the `.cache/shards` directory will prevent the down of the shards which should be the main source of time penalty.

In scenarios where a new version of the dependecy is released, caching the `$TRAVIS_BUILD_DIR/lib` will prevent checking the new release.

Reply:
I guess you're right. When using a `shard.lock`, this shouldn't matter, though. But it's probably okay to cache only the shards cache, not checked out lib.